### DISCUSSION TITLE
Support Vercel middleware rewrites

### DISCUSSION THREAD

**Discussion by alexanderflink** [withastro/roadmap#261] [Proposal]
Hello!

Currently on Next.js for all our ecommerce projects, however Astro is looking really good. One major feature that we love with Vercel + Next.js currently is combining ISR with middleware rewrites. This allows us to statically render pages with "dynamic" content such as different pricing based on the users market (which is stored in a cookie). I think Vercel manages this by sending an `x-middleware-rewrite` header to the Next.js app when you do `NextResponse.rewrite()` from their middleware functions. Would it be possible for Astro to listen to these kinds of headers (or to implement that logic yourself) to rewrite the actual Astro page being rendered, while leaving the URL intact?

This would allow us to use Astro SSG, while still having very dynamic pages. We would basically be able to statically generate every combination of a page possible, and rewrite to it from a Vercel middleware function based on user information.

**Comment by anuragk15** [comment]
Sad to see that this feature is not yet available. The middleware functions and rewrites are also critical multi tenant apps where you have multiple custom domains

**Comment by florian-lefebvre** [comment]
I'm also interested in this for i18n specific 404 pages

**Comment by Andrew-Chen-Wang** [comment]
It would be great to have this. My typical stack looks like an SSR framework like NextJS / Astro, plain React as an SPA for authenticated users, and a Python backend acting as the main backend. The SSR framework just calls APIs to the main backend; the main backend acts as a kind of API Gateway.

This would be great to have for multi tenant purposes or in general personal projects. It would dramatically save personal project costs for people who have this kind of set up / despise JS backend development.
